### DISCUSSION TITLE
OpenSSL provider implementation: custom functions are not called

### DISCUSSION THREAD

**Discussion by commel** [openssl/openssl#23549] [Q&A] [answered]
I am working on a custom provider. I've registered the init method and the provider is loaded as well. But somehow I cannot get my custom signature functions to be called. I can see the op codes being worked through, but after that none of my custom functions are being executed, neither KeyManagement, Store, Signature,....

Also I need the default provider to load the private and public keys, I do not want to override that behavior here, it is loaded in second place, like `-provider mycustom -provider default`. To specify which provider to use, I added `-propquery provider=mycustom`.

The dispatch registration for any of these op codes (KeyManagement, Store, Signature) defines `ED25519:1.3.101.112` and the PropQuery `provider=mycustom`:

```
static const OSSL_ALGORITHM ed25519ph_ḱeymgmt_dispatch[] = {
    {"ED25519:1.3.101.112", "provider=mycustom", ed25519ph_keymgmt_functions, PROV_DESCS_ED25519},
    {NULL, NULL, NULL, NULL}};
```
The functions are registered like this:
```
const OSSL_DISPATCH ed25519ph_keymgmt_functions[] = {
    {OSSL_FUNC_KEYMGMT_NEW, (void (*)(void)) ed25519ph_keymgmt_new},
    {OSSL_FUNC_KEYMGMT_FREE, (void (*)(void)) ed25519ph_keymgmt_free},    
    {OSSL_FUNC_KEYMGMT_LOAD, (void (*)(void)) ed25519ph_keymgmt_load},
    {OSSL_FUNC_KEYMGMT_HAS, (void (*)(void)) ed25519ph_keymgmt_has},
    OSSL_DISPATCH_END};
```
(the others are similar)

And just to see, if the method is called, I've implemented a stub:
```
static void* ed25519ph_keymgmt_new(void *provctx) 
{
  fprintf(stdout, "%s - keymgmt new", __FILE__);
  struct provider_ctx_st *pctx = provctx;

  EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new(NULL, NULL);
  pctx->pkctx = ctx;
  return ctx;
}
```

In the end I try to run it:
```
openssl pkeyutl -provider-path build/ -provider mycustom -provider default -propquery ?provider=mycustom -verify -inkey ${PUBLIC} -pubin -in ${MESSAGE} -rawin -sigfile ${SIGNATURE}	
```

This is the result:
```
provider init...
new provider context
operating switch: 22
operating switch: 21
operating switch: 10
key management dispatch: 0x7fe4c7377700
operating switch: 21
operating switch: 21
operating switch: 12
operating switch: 1
Signature Verification Failure
teardown
free provider context
```

The "operating switch" print out is from my OSSL_ALGORITHM-method I've registered for the OSSL_FUNC_PROVIDER_QUERY_OPERATION.

I have no ideas left, why my method would not be called. Hopefully someone can hint me on what I am missing here.

**Comment by commel** [comment]
I can call the methods explicitly, so somehow I screwed up the lookup mechanism. Do I need to define the "provider"-parameter in the provider?

Proof, that the method call works from within the dispatch table:
```
  case OSSL_OP_SIGNATURE: // 12
    fprintf(stdout, "signature dispatch: %p\n", (void*) ed25519ph_signature_dispatch[0].implementation[0].function);
    ed25519ph_signature_dispatch[0].implementation[0].function();
```
output:
```
operating switch: 12
signature dispatch: 0x7fe0c4c31640
ed25519ph_sign.c - digest sign newctx

```

**Comment by mattcaswell** [comment]
The key is being loaded by the decoder in the default provider (because you don't have any decoders in your provider). Once loaded it then tries to find a suitable key manager. This will prefer the key manager from your provider due to the "?provider=mycustom" that you have passed as a property query. However in order to pass the key from the default provider and into yours, libcrypto will attempt to "export" it from default and "import" it into your provider. Unfortunately you have not implemented the "import"/"import_types" functions in your key manager so this fails. libcrypto will then fallback to the key manager in the default provider - and use the default provider from there on.

In other words you must implement the "import"/"import_types" functions in your key manager.

**Reply by commel** [reply]
Thanks a lot, I will try this out! Sounds logical!

**Reply by commel** [reply]
Thanks again for your insights. I've extended the keymanagement implementation like this:
```
static OSSL_FUNC_keymgmt_import_fn ed2559ph_keymgmt_import;
static OSSL_FUNC_keymgmt_import_types_fn ed2559ph_keymgmt_import_types;
static OSSL_FUNC_keymgmt_import_types_ex_fn ed2559ph_keymgmt_import_types_ex;

{OSSL_FUNC_KEYMGMT_IMPORT, (void (*)(void)) ed2559ph_keymgmt_import},
{OSSL_FUNC_KEYMGMT_IMPORT_TYPES, (void (*)(void)) ed2559ph_keymgmt_import_types},
{OSSL_FUNC_KEYMGMT_IMPORT_TYPES_EX, (void (*)(void)) ed2559ph_keymgmt_import_types_ex},
```

The log shows that openssl tries to use the operation 10 (keymgmt) but it still wont execute my provider's functions:
```
operating switch: 22
operating switch: 21
operating switch: 10
operating switch: 21
operating switch: 21
operating switch: 10
operating switch: 12
operating switch: 1
```
Does the mapping not fit here maybe?
```
#define PROV_NAMES_ED25519 "ED25519:1.3.101.112:Ed25519:ed25519:Ed25519ph:ed25519ph:ED25519ph:1.3.101.114"

static const OSSL_ALGORITHM ed25519ph_ḱeymgmt_dispatch[] = {
    {PROV_NAMES_ED25519, "provider=mycustom", ed25519ph_keymgmt_functions, PROV_DESCS_ED25519},
    {NULL, NULL, NULL, NULL}};
```

BTW, after the op-handling i get these error messages, are they a hint for something I am missing here?
```
40CCCB524E7F0000:error:030000C5:digital envelope routines:geterr:cannot get parameters:crypto/evp/evp_utils.c:65:
40CCCB524E7F0000:error:030000E1:digital envelope routines:evp_md_from_algorithm:cache constants failed:crypto/evp/digest.c:1129:
```

Is there a way to see, what the requests are specifically?

**Reply by commel** [reply]
In the operations method ```const OSSL_ALGORITHM *ed25519ph_operation(void *vprovctx, int operation_id, int *no_store)``` I've always set ```*no_store = 0```, but when I changed it to 1, at least the ```OSSL_FUNC_KEYMGMT_NEW``` method is called. But still not the import-Methods.

**Reply by commel** [reply, marked as answer]
Turns out I needed to provide a signature context, even if it just was an empty byte pointer.

**Reply by iomartin** [reply]
@commel I'm going through a very similar issue (except I don't have my own keymgmt), where did you set this signature context?

**Reply by commel** [reply]
> @commel I'm going through a very similar issue (except I don't have my own keymgmt), where did you set this signature context?

You need to implement functions matching `OSSL_FUNC_signature_newctx_fn` and `OSSL_FUNC_signature_freectx_fn` and return some struct to hold you context. If the context is empty this is fine, but it must not be NULL. See https://www.openssl.org/docs/man3.0/man7/provider-signature.html

**Reply by iomartin** [reply]
Thanks for the reply, I already have a newctx, freectx, sign_init and sign functions. But none even get called. I can see my `QUERY_OPERATION` being called, though. Below is a simplified version of the provider and a sample output run.

```C
#include <openssl/provider.h>
#include <openssl/core_dispatch.h>

#include <stdio.h>

static OSSL_FUNC_signature_newctx_fn newctx;
static OSSL_FUNC_signature_freectx_fn freectx;
static OSSL_FUNC_signature_sign_init_fn ecdsa_sign_init;
static OSSL_FUNC_signature_sign_fn ecdsa_sign;

static void *newctx(void *provctx, const char *propq)
{
	printf("%s\n", __func__);
	return (void *) -1;
}

static void freectx(void *vctx)
{
	printf("%s\n", __func__);
}

static int sign_init(void *vctx, void *ec, const OSSL_PARAM params[])
{
        printf("%s\n", __func__);
	return 0;
}

static int sign(void *vctx, unsigned char *sig, size_t *siglen,
                      size_t sigsize, const unsigned char *tbs, size_t tbslen)
{
	printf("%s\n", __func__);
	return 0;
}

static const OSSL_DISPATCH myprov_ecdsa_sig_functions[] = {
    { OSSL_FUNC_SIGNATURE_NEWCTX, (void (*)(void))newctx },
    { OSSL_FUNC_SIGNATURE_SIGN_INIT, (void (*)(void))sign_init },
    { OSSL_FUNC_SIGNATURE_SIGN, (void (*)(void))sign },
    { OSSL_FUNC_SIGNATURE_FREECTX, (void (*)(void))freectx },
    { 0, NULL },
};

static const OSSL_ALGORITHM myprov_sig_list[] = {
	{ "ECDSA", NULL, myprov_ecdsa_sig_functions },
	{ NULL, NULL, NULL },
};

static const OSSL_ALGORITHM *myprov_query(void *provctx, int operation_id,
					    int *no_cache)
{
	*no_cache = 1;

	printf("%s operation_id=%d\n", __func__, operation_id);

	switch (operation_id) {
	case OSSL_OP_SIGNATURE:
		return myprov_sig_list;
	}

	return NULL;
}

static void myprov_teardown(void *provctx)
{
	printf("%s\n", __func__);
}

static int myprov_get_params(void *provctx, OSSL_PARAM params[])
{
	printf("%s\n", __func__);
	return 1;
}

static const OSSL_PARAM *myprov_gettable_params(void *provctx)
{
	printf("%s\n", __func__);
	return NULL;
}

static const OSSL_DISPATCH myprov_functions[] = {
    { OSSL_FUNC_PROVIDER_TEARDOWN, (void (*)(void)) myprov_teardown },
    { OSSL_FUNC_PROVIDER_QUERY_OPERATION, (void (*)(void)) myprov_query },
    { OSSL_FUNC_PROVIDER_GET_PARAMS, (void (*)(void)) myprov_get_params },
    { OSSL_FUNC_PROVIDER_GETTABLE_PARAMS, (void (*)(void)) myprov_gettable_params },
    { 0, NULL }
};

int OSSL_provider_init(const OSSL_CORE_HANDLE *core,
		       const OSSL_DISPATCH *in,
		       const OSSL_DISPATCH **out,
		       void **vprovctx)
{
	printf("%s\n", __func__);
	*out = myprov_functions;
	*vprovctx = NULL;
	return 1;
}
````

```
$ openssl dgst -provider-path . -provider myprov -provider default -sign ec256r1.priv -out sig.out file
OSSL_provider_init
myprov_query operation_id=1
myprov_query operation_id=22
myprov_query operation_id=21
myprov_query operation_id=10
myprov_query operation_id=21
myprov_query operation_id=21
myprov_query operation_id=12
myprov_query operation_id=10
myprov_query operation_id=5
myprov_query operation_id=2
myprov_teardown

```

**Reply by commel** [reply]
Yeah, you need a keymgmt implementation, thats a requirement for a signature implementation. I've fallen for the same trap.

**Reply by iomartin** [reply]
Oooh right. I missed this part from https://www.openssl.org/docs/man3.0/man7/provider-signature.html:

> A signature algorithm must also implement some mechanism for generating, loading or importing keys via the key management (OSSL_OP_KEYMGMT) operation. See [provider-keymgmt(7)](https://www.openssl.org/docs/man3.0/man7/provider-keymgmt.html) for further details.

I'll try that next

**Reply by iomartin** [reply]
@commel I'm going around in circles...

I'm trying to avoid implementing all of the keymgmt myself, I'm hoping to just let the default provider handle everything. But I can't seem to find a way to call it. I tried doing something like this for `OSSL_FUNC_KEYMGMT_NEW` for example:

```C
static void *keymgmt_new(void *provctx)
{
        EVP_KEYMGMT *keymgmt = EVP_KEYMGMT_fetch(NULL, "EC", "provider=default");
        return evp_keymgmt_newdata(keymgmt);
}
```

But `evp_keymgmt_newdata()` is an internal API and I can't find another way to call the default provider.

I also tried adding this to my `QUERY_OPERATION`:

```C
static const OSSL_ALGORITHM *kpt_prov_query(void *provctx, int operation_id, int *no_cache)
{
        OSSL_PROVIDER *dflt = OSSL_PROVIDER_load(NULL, "default");
        switch(operation_id) {
        case OSSL_OP_KEYMGMT:
                return OSSL_PROVIDER_query_operation(dflt, operation_id, no_cache);
        case OSSL_OP_SIGNATURE:
                return myprov_sig_list;
        }

        return NULL
}
```

But with that it still doesn't call any of my signature functions.

Any ideas?

**Reply by commel** [reply]
You have to implement it yourself, but internally you can call default methods. Anyway when initiating the keymgmt_new you need to return a context as well. An empty struct maybe, but something. See https://www.openssl.org/docs/man3.0/man7/provider-keymgmt.html

Also handle the keymgmt query operation similar to your signature operation.

In keymgmt you will receive params that contain your key, in your case you might want to build and EVP_PKEY from it. That would happen in `OSSL_FUNC_keymgmt_import_fn`. There you can load it via default methods and return the EVP_PKEY.

Watch out to implement all the pair methods mentioned in the docs. If one of these is missing, the manager will go right over it. Here a pair is the new and free-methods.

**Reply by iomartin** [reply]
Thanks @commel!!! It's finally calling my functions!

For reference if anyone stumbles on this in the future, here's a list of functions that I needed to define:

Signature:
OSSL_FUNC_SIGNATURE_NEWCTX
OSSL_FUNC_SIGNATURE_FREECTX
OSSL_FUNC_SIGNATURE_SIGN_INIT
OSSL_FUNC_SIGNATURE_SIGN
OSSL_FUNC_SIGNATURE_DIGEST_SIGN_INIT
OSSL_FUNC_SIGNATURE_DIGEST_SIGN_UPDATE
OSSL_FUNC_SIGNATURE_DIGEST_SIGN_FINAL
OSSL_FUNC_SIGNATURE_DUPCTX

Keymgmt:
OSSL_FUNC_KEYMGMT_NEW
OSSL_FUNC_KEYMGMT_FREE
OSSL_FUNC_KEYMGMT_HAS
OSSL_FUNC_KEYMGMT_IMPORT
OSSL_FUNC_KEYMGMT_IMPORT_TYPES

These were sufficient to have my sign functions called, I haven't started verify yet, so there might be a few others (besides the obvious OSSL_FUNC_SIGNATURE_*VERIFY*)

**Reply by mattcaswell** [reply]
@iomartin - it would be great if you could make a PR with any improvements in our documentation that you can suggest based on what you've learned.
